contract/src/lib.rs

```diff
@@ -290,10 +293,32 @@ impl MpcContract {
     }
 
     #[allow(unused_variables)]
-    pub fn sign(&mut self, payload: [u8; 32]) -> [u8; 32] {
-        near_sdk::env::random_seed_array()
+    pub fn sign(&mut self, payload: [u8; 32]) -> Promise {
+        let hash = env::sha256(&payload);
+        self.pending_requests.insert(&hash, &None);
+        Self::ext(env::current_account_id()).sign_helper(hash)
+    }
+
+    pub fn sign_helper(&mut self, payload_id: Vec<u8>) -> PromiseOrValue<String> {
+        if let Some(signature) = self.pending_requests.get(&payload_id) {
+            match signature {
+                Some(signature) => {
+                    self.pending_requests.remove(&payload_id);
+                    PromiseOrValue::Value(signature)
+                }
+                None => {
+                    env::log_str("not ready");
```

Comment:
logging so that we can easily see how many times it is able to recurse

Reply:
should we also add a counter to keep track of the recurse chain more easily?